File: src/commonMain/kotlin/com/pricecompare/model/PurchaseContext.kt
package com.pricecompare.model

/**
 * src/commonMain/kotlin/com/pricecompare/model/PurchaseContext.kt
 *
 * 購入コンテキストを表すデータクラス。
 * 商品単位の情報を超える、注文全体の情報を含む。
 * 送料、注文クーポン、ポイント情報を管理する。
 *
 * 関連ファイル:
 * - src/commonMain/kotlin/com/pricecompare/model/Money.kt
 * - src/commonMain/kotlin/com/pricecompare/model/Discount.kt
 * - src/commonMain/kotlin/com/pricecompare/pricing/PriceCalculator.kt
 */
data class PurchaseContext(
    /** 送料（null=未入力） */
    val shippingCost: Money? = null,

    /** 送料無料条件（この金額以上で送料無料） */
    val freeShippingThreshold: Money? = null,

    /** 注文全体に適用されるクーポン */
    val orderCoupons: List<Discount.CouponDiscount> = emptyList(),

    /** 使用ポイント */
    val usedPoints: Int = 0,

    /** 獲得予定ポイント */
    val earnedPoints: Int = 0,

    /** ポイント評価率（1.0 = 1ポイント1円） */
    val pointEvaluationRate: Decimal = Decimal.ONE
) {
    /**
     * 送料無料条件を満たすかどうかを判定する。
     * クーポン適用後・ポイント使用前の金額で判定。
     */
    fun isShippingFree(subtotalAfterCoupons: Money): Boolean {
        val threshold = freeShippingThreshold ?: return false
        return subtotalAfterCoupons.amount.compareTo(threshold.amount) >= 0
    }

    /**
     * 実送料を計算する。
     * 送料未入力の場合はnullを返す。
     */
    fun effectiveShipping(subtotalAfterCoupons: Money): Money? {
        if (shippingCost == null) return null
        if (isShippingFree(subtotalAfterCoupons)) return Money.ZERO
        return shippingCost
    }

    /**
     * 獲得ポイントの評価額を計算する。
     */
    fun earnedPointsValue(): Money =
        Money(Decimal.fromLong(earnedPoints.toLong()) * pointEvaluationRate)

    /**
     * 使用ポイントの評価額を計算する。
     */
    fun usedPointsValue(): Money =
        Money(Decimal.fromLong(usedPoints.toLong()) * pointEvaluationRate)
}

File: flutter_app/lib/engine/purchase_context.dart
import 'decimal.dart';
import 'money.dart';
import 'offer.dart';

class PurchaseContext {
  final Money? shippingCost;
  final Money? freeShippingThreshold;
  final List<CouponDiscount> orderCoupons;
  final int usedPoints;
  final int earnedPoints;
  final Decimal pointEvaluationRate;

  PurchaseContext({
    this.shippingCost,
    this.freeShippingThreshold,
    this.orderCoupons = const [],
    this.usedPoints = 0,
    this.earnedPoints = 0,
    Decimal? pointEvaluationRate,
  }) : pointEvaluationRate = pointEvaluationRate ?? Decimal.one {
    if (usedPoints < 0 || earnedPoints < 0) {
      throw ArgumentError('Points must be non-negative');
    }
    if (this.pointEvaluationRate < Decimal.zero ||
        this.pointEvaluationRate > Decimal('2')) {
      throw ArgumentError.value(
        this.pointEvaluationRate,
        'pointEvaluationRate',
        'must be between 0 and 2',
      );
    }
  }

  bool isShippingFree(Money subtotalAfterCoupons) {
    final threshold = freeShippingThreshold;
    return threshold != null && subtotalAfterCoupons.amount >= threshold.amount;
  }

  Money? effectiveShipping(Money subtotalAfterCoupons) {
    final shipping = shippingCost;
    if (shipping == null) {
      return null;
    }
    return isShippingFree(subtotalAfterCoupons) ? Money.zero : shipping;
  }

  Money earnedPointsValue() =>
      Money(Decimal.fromInt(earnedPoints) * pointEvaluationRate);

  Money usedPointsValue() =>
      Money(Decimal.fromInt(usedPoints) * pointEvaluationRate);
}
